src/replica/HttpAsyncClient.cc

```diff
@@ -0,0 +1,399 @@
+            _expirationTimer.async_wait(bind(&HttpAsyncClient::_expired, shared_from_this(), _1));
+        }
+        _state = State::IN_PROGRESS;
+    } catch(exception const& ex) {
+        _finish(lock, State::FAILED, ex.what());
+        throw;
+    }
+}
+
+
+void HttpAsyncClient::cancel() {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    switch (_state) {
+        case State::CREATED:
+            _finish(lock, State::CANCELLED);
+            return;
+        case State::IN_PROGRESS:
+            _finish(lock, State::CANCELLED);
+            return;
+        default:
+            // Nothing to be done in the final states of the request.
+            break;
+    }
+}
+
+
+string HttpAsyncClient::error() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    return _error;
+}
+
+
+int HttpAsyncClient::responseCode() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED, State::BODY_LIMIT_ERROR});
+    auto const& header = _res.get().base();
+    return header.result_int();
+}
+
+
+map<string, string> const& HttpAsyncClient::responseHeader() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED, State::BODY_LIMIT_ERROR});
+    return _responseHeader;
+}
+
+
+string const& HttpAsyncClient::responseBody() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED});
+    return _res.get().body();
+}
+
+
+size_t HttpAsyncClient::responseBodySize() const {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::FINISHED});
+    return _res.get().body().size();
+}
+
+
+void HttpAsyncClient::_parsUrl(util::Lock const& lock) {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    regex re("^[ ]*http://([^:^/]+)?(:([0-9]+))?(/([^ ]*))?[ ]*$", regex::extended);
+    smatch match;
+    if (!regex_search(_url, match, re)) {
+        throw invalid_argument(context + "incorrect syntax of the URL '" + _url + "'.");
+    }
+    if (match.size() != 6) {
+        throw invalid_argument(context + "problem with the regular expression when parsing URL '" + _url + "'.");
+    }
+    auto const parseVal = [](string const& in, string& out) {
+        if (!in.empty()) out = in;
+    };
+    parseVal(match[1].str(), _host);
+    parseVal(match[3].str(), _port);
+    parseVal(match[4].str(), _target);
+}
+
+
+void HttpAsyncClient::_prepareRequest(util::Lock const& lock) {
+    _req.version(11);   // HTTP/1.1
```

Comment:
Grrr. You should definitely switch to Python.

Reply:
Well, `11` is what the `Boost.Beast` expects from me. I wish we had (in C++) a good library that would be as convenient as Python's `requests` (AKA "HTTP for humans"). In theory, `Boost.Beast` provides a good fundation for building such library. However, this was not my intent here as developing such universally useful high-level library would take a much big effort.